pkg/compactor/blocks_cleaner.go

```diff
@@ -223,10 +225,72 @@ func (c *BlocksCleaner) deleteUser(ctx context.Context, userID string) error {
 	c.tenantBlocks.DeleteLabelValues(userID)
 	c.tenantMarkedBlocks.DeleteLabelValues(userID)
 
-	level.Info(userLogger).Log("msg", "finished deleting blocks for user marked for deletion", "deletedBlocks", deleted)
+	if deleted > 0 {
+		level.Info(userLogger).Log("msg", "deleted blocks for user marked for deletion", "deletedBlocks", deleted)
+	} else {
+		level.Info(userLogger).Log("msg", "no blocks found")
+	}
+
+	mark, err := cortex_tsdb.ReadTenantDeletionMark(ctx, c.bucketClient, userID)
+	if err != nil {
+		return errors.Wrap(err, "failed to read tenant deletion mark")
+	}
+	if mark == nil {
+		return errors.Wrap(err, "cannot find tenant deletion mark anymore")
+	}
+
+	// If we have just deleted some blocks, update "finished" time. Also update "finished" time if it wasn't set yet, but there are no blocks.
+	// Note: this UPDATES the tenant deletion mark. Components that use caching bucket will NOT SEE this update,
```

Comment:
Any chance of multiple callers updating it? I assume if they do they will write similar times and it will be ok?

Reply:
Within one compactor, it is only called once.

Multiple compactors may happen to call this method concurrently (eg. one compactor is doing compaction, and another compactor starts and "takes over" the user). In that case, they may overwrite each-others timestamp if they both happen to delete some blocks or read tenant deletion mark with no timestamp. In that case last write will win, which will delay eventual final cleanup of data for tenant.